include/deal.II/base/mpi.h

```diff
@@ -547,6 +564,37 @@ namespace Utilities
           }
       }
 
+      template <typename T>
+      inline
+      void all_reduce (const MPI_Op    &mpi_op,
+                       const Vector<T> &values,
+                       const MPI_Comm  &mpi_communicator,
+                       Vector<T>  &output)
+      {
+#ifdef DEAL_II_WITH_MPI
+        if (job_supports_mpi())
+          {
+            if (values.begin() != output.begin())
+              output.reinit (values.size());
+
+            MPI_Allreduce ((values.begin() != output.begin()
+                            ?
+                            const_cast<void *>(static_cast<const void *>(values.begin()))
+                            :
+                            MPI_IN_PLACE),
+                           output.begin(), values.size(), internal::mpi_type_id((T *)0), mpi_op,
+                           mpi_communicator);
+          }
+        else
+#endif
+          {
+            (void)mpi_op;
+            (void)mpi_communicator;
+            output = values;
+          }
+      }
+
```

Comment:
I'll wait for other opinion. I don't feel strongly about it, but I always like to reduce the number of code lines.

Regarding the non-MPI cases, the loop from 0 to N fits all cases (the variant right now used for the fixed-sized array which would get variable-sized of course). Resizing is more subtle I agree. As you say, it could be pulled out if we wanted.